src/runner_manager.py

```diff
@@ -0,0 +1,358 @@
+        if self.proxies:
+            # setup proxy for requests
+            self.session.proxies.update(self.proxies)
+            # add proxy to fastcore which ghapi uses
+            proxy = urllib.request.ProxyHandler(self.proxies)
+            opener = urllib.request.build_opener(proxy)
+            fastcore.net._opener = opener
+
+        self._github = GhApi(token=token)
+        self._jinja = jinja2.Environment(
+            loader=jinja2.FileSystemLoader("templates"), autoescape=True
+        )
+        self._lxd = pylxd.Client()
+
+        if "/" in path:
+            owner, repo = path.split("/")
+            self.path = GitHubRepo(owner=owner, repo=repo)
+        else:
+            self.path = GitHubOrg(org=path)
+
+        self.app_name = app_name
+        self.reconcile_interval = reconcile_interval
+        self.image = image
+
+    @classmethod
+    @retry(tries=10, delay=15, max_delay=60, backoff=1.5)
+    def install_deps(cls) -> None:
+        """Install dependencies."""
+
+        logger.info("Installing charm dependencies.")
+
+        # Binding for snap, apt, and lxd init commands are not available so subprocess.run used.
+        execute_command(["/usr/bin/apt", "remove", "-qy", "lxd", "lxd-client"], check=False)
+        execute_command(["/usr/bin/snap", "install", "lxd", "--channel=latest/stable"])
+        execute_command(["/usr/bin/snap", "refresh", "lxd", "--channel=latest/stable"])
+        execute_command(["/snap/bin/lxd", "waitready"])
+        execute_command(["/snap/bin/lxd", "init", "--auto"])
+        execute_command(["/usr/bin/chmod", "a+wr", "/var/snap/lxd/common/lxd/unix.socket"])
+        execute_command(["/snap/bin/lxc", "network", "set", "lxdbr0", "ipv6.address", "none"])
+        execute_command(
+            [
+                "/usr/bin/apt",
+                "install",
+                "-qy",
+                "cpu-checker",
+                "libvirt-clients",
+                "libvirt-daemon-driver-qemu",
+            ],
+        )
+        cls.runner_bin_path.parent.mkdir(parents=True, exist_ok=True)
+
+        logger.info("Finished installing charm dependencies.")
+
+    def get_latest_runner_bin_url(self, os_name: str = "linux", arch_name: str = "x64") -> str:
+        """Get the URL for the latest runner binary.
+
+        The runner binary URL changes when a new version is available.
+
+        Returns:
+            The URL to the latest runner binary.
+        """
+
+        runner_bins: RunnerApplicationList = []
+        if isinstance(self.path, GitHubRepo):
+            runner_bins = self._github.actions.list_runner_applications_for_repo(
+                owner=self.path.owner, repo=self.path.repo
+            )
+        elif isinstance(self.path, GitHubOrg):
+            runner_bins = self._github.actions.list_runner_application_for_org(org=self.path.org)
+
+        logger.debug("Response of runner binary list: %s", runner_bins)
+
+        for bin in runner_bins:
```

Comment:
there is a nice trick using `next`, e.g., `next(binary.download_url for binary in runner_bins if binary.os == os_name and bin.architecture == arch_name)`, you can give a default or handle the exception (maybe `StopIteration`? in case nothing is found. I was also wondering, would there ever be multiple versions available for a combination of os_name and architecture? Or only one?

Reply:
I checked API endpoint again.
It seems to only return one version of os/arch, and it is returning the `sha256_checksum` (I thought it did not).
Although these two are not stated by the API. 
`sha256_checksum` is a optional field in the API, and nothing about one version per os/arch is mentioned.

Reply:
it should be ok regarding the os/arch, I'm guessing the API just returns the latest release. I was thinking it might return binaries from earlier releases which is why I asked the question